### PR TITLE
managedJobsNamespaceSelector for Deployment, StatefulSets, and Pods

### PR DESCRIPTION
**Pull request by dgrove-oss**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Normalizes handling of manageJobsWithoutQueueName and managedJobsNamespaceSelector for the Pods, Deployment, and StatefulSet integrations.

#### Which issue(s) this PR fixes:

Part of #3589 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/jobs/pod/pod_webhook.go [hunk 1]
```diff
@@ -209,10 +209,11 @@ func (w *PodWebhook) Default(ctx context.Context, obj runtime.Object) error {
 				return err
 			}
 		}
+
+		// copy back to the object
+		pod.pod.DeepCopyInto(obj.(*corev1.Pod))
```
Comment (dgrove-oss) [inline pkg/controller/jobs/pod/pod_webhook.go:214]: Moving the copy inside the that actually mutates the pod is a small cleanup / micro-optimization.  I could revert if we want to do it unconditionally (in case there is a future change that introduces another branch that also mutates the pod).

Comment (mimowo) [reply pkg/controller/jobs/pod/pod_webhook.go:214]: Good catch, but I suggest a separate PR. We may consider backporting it to previous releases as a performance fix. Also, in case there is a bug - we can rollback easily just that.

File: pkg/controller/jobs/statefulset/statefulset_webhook.go [hunk 2]
```diff
@@ -64,23 +69,34 @@ func (wh *Webhook) Default(ctx context.Context, obj runtime.Object) error {
...
 
 	queueName := jobframework.QueueNameForObject(ss.Object())
-	if queueName == "" {
-		return nil
-	}
 
-	if ss.Spec.Template.Labels == nil {
-		ss.Spec.Template.Labels = make(map[string]string, 2)
+	// Do not manage a statefulset without a queue name unless the namespace selector also matches
+	if features.Enabled(features.ManagedJobsNamespaceSelector) && wh.manageJobsWithoutQueueName && queueName == "" {
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_webhook.go:74]: I think this is a bit hard to follow, espacially because `manageJobsWithoutQueueName` and `queueName` are checked in both `if`s - here and below. The code is also duplicated with Jobs.

So, I would suggest to commonize the code, maybe like that:
```golang
func WorkloadManagedByKueue(ctx context.Context, jobObj client.Object, k8sClient client.Client,
	manageJobsWithoutQueueName bool, managedJobsNamespaceSelector labels.Selector) (bool, error) {
	// If the owner is managed, then the workload is also managed
	if owner := metav1.GetControllerOf(jobObj); owner != nil && IsOwnerManagedByKueue(owner) {
	   return true, nil
	}
	// we manage Jobs with queue name
	if QueueNameForObject(jobObj) != "" { 
	     return true, nil
	}
	// Manage Jobs if manageJobsWithoutQueueName and namespace selector matches
	if features.Enabled(features.ManagedJobsNamespaceSelector) && manageJobsWithoutQueueName {
		ns := corev1.Namespace{}
		err := k8sClient.Get(ctx, client.ObjectKey{Name: jobObj.GetNamespace()}, &ns)
		if err != nil {
			return false, fmt.Errorf("failed to get namespace: %w", err)
		}
		if managedJobsNamespaceSelector.Matches(labels.Set(ns.GetLabels())) {
			return true, nil
		}
	}
	return false, nil
}
```
I know this is more complex for pods, so we could commonize with pods at another chance. If there is more work than on the surface I'm ok with a follow up.

**Review by mimowo — COMMENTED** [review]
LGTM overall. but I would like to commonize the code and add some tests (either unit or integration). If you commonize the code with Jobs I believe the tests could be a follow up as there would be little duplication.

File: pkg/controller/jobs/statefulset/statefulset_webhook.go (hunk 2 above)
Comment (dgrove-oss) [reply pkg/controller/jobs/statefulset/statefulset_webhook.go]: Good suggestion.  It was straightforward to pull out helper method and make it easier to see what is happening.

**Review by mimowo — COMMENTED** [review]
/lgtm
/approve
Inifially I thought we should filter at the higher-level (STS and Deployment), but I don't see any issues delegating the filtering to the pod level.

In any case, the logic around "queue-name" is getting complex, so I think we need to strenghten our testing. We already have e2e tests for `manageJobsWithoutQueueName: false`, but they mostly focus on the mechanics of the integration. I hope things like queue-name logic can be covered in integration tests (with webhooks enabled) (opened: https://github.com/kubernetes-sigs/kueue/issues/3767)
